### PR TITLE
[MLIR][SPIRV] Add definition and (de)serialization for cache controls

### PR DESCRIPTION
**Pull request by victor-eds**
[SPV_INTEL_cache_controls](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_cache_controls.html) defines decorations for load and store cache control. Add support for this extension in the SPIR-V dialect.

As several `CacheControlLoadINTEL` and `CacheControlStoreINTEL` may be applied to the same value, these are represented as array attributes. (De)Serialization takes care of this representation.

### CODE REVIEW

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp [hunk 1]
```diff
@@ -226,6 +226,28 @@ spirv::Deserializer::processMemoryModel(ArrayRef<uint32_t> operands) {
...
+           << decorationName << "needs a cache control integer literal and a "
+           << cacheControlKind << " cache control literal";
+  }
+  unsigned cacheLevel = words[2];
+  auto cacheControlAttr = static_cast<EnumTy>(words[3]);
+  auto value = opBuilder.getAttr<AttrTy>(cacheLevel, cacheControlAttr);
+  SmallVector<Attribute> attrs;
+  if (auto attrList =
+          llvm::dyn_cast_or_null<ArrayAttr>(decorations[words[0]].get(symbol)))
+    attrs.append(attrList.begin(), attrList.end());
```
Comment (kuhar) [inline mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:245]: nit: you can use `llvm::append_range(attrs, attrList);`. either way is fine, just FYI

File: mlir/lib/Target/SPIRV/Serialization/Serializer.cpp [hunk 2]
```diff
@@ -217,10 +217,43 @@ static std::string getDecorationName(StringRef attrName) {
...
+    auto cacheControlAttr = dyn_cast<AttrTy>(attr);
+    if (!cacheControlAttr) {
+      return emitError(loc, "expecting array attribute of ")
+             << attrName << " for " << stringifyDecoration(decoration);
+    }
+    // This named attribute encodes several decorations. Emit one per
+    // element in the array.
+    LogicalResult res = emitter(cacheControlAttr);
+    if (failed(res))
+      return res;
```
Comment (kuhar) [inline mlir/lib/Target/SPIRV/Serialization/Serializer.cpp:252]: nit
```suggestion
    if (failed(emitter(cacheControlAttr)))
      return failure();
```

**Review by kuhar — COMMENTED** [review]
Looks good overall. Could we add some tests to cover the failure cases? Not sure how easy that'd be.

File: mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp]: I always forget about the name. Thanks!

**Review by victor-eds — COMMENTED** [review]
Thanks for the comments, @kuhar. I added failure tests covering encoding. I couldn't see similar tests for decoding, so I assume that isn't covered and input is assumed to be valid SPIR-V.

**Review by kuhar — APPROVED** [review]

**Comment by victor-eds** [comment]
I'd like to get this merged today if there aren't additional comments from reviewers.

**Pull request merged**